File: examples/python/pay.py
"""Shared setup for the Python examples: an httpx client that pays x402 challenges from your
wallet, only on the networks you allow and never above $0.10 a call.

Set CROSSCHECK_NETWORKS=eip155:84532 to pay with free test USDC on Base Sepolia.
"""

import asyncio
import os

import httpx
from eth_account import Account
from x402 import max_amount, prefer_network, x402Client
from x402.http.clients import x402HttpxClient

from x402.mechanisms.evm.exact.register import register_exact_evm_client

NETWORKS = [n.strip() for n in os.environ.get("CROSSCHECK_NETWORKS", "eip155:8453").split(",")]


def paying_client() -> x402HttpxClient:
    payer = x402Client()
    register_exact_evm_client(payer, Account.from_key(os.environ["CROSSCHECK_WALLET_KEY"]), networks=NETWORKS)
    for network in reversed(NETWORKS):
        payer.register_policy(prefer_network(network))
    payer.register_policy(max_amount(100_000))  # never more than $0.10 (USDC has 6 decimals)
    return x402HttpxClient(payer, timeout=120)


async def post_and_wait(url: str, body: dict) -> dict:
    """POST a paid request and wait for the result (a 202 means paid and queued)."""
    async with paying_client() as http:
        res = await http.post(url, json=body)
    out = res.json()
    if res.status_code not in (200, 202):
        raise RuntimeError(f"crosscheck returned {res.status_code}: {out}")
    result_url, token = out.get("result_url"), out.get("result_token")
    async with httpx.AsyncClient(timeout=30) as plain:
        while out.get("status") == "pending":
            await asyncio.sleep(out.get("retry_after_seconds", 5))
            out = (await plain.get(result_url, headers={"authorization": f"Bearer {token}"})).json()
    return out

File: examples/typescript/pay.ts
// Shared setup for the examples: a fetch that pays x402 challenges from your wallet,
// only on the networks you allow and never above $0.10 a call.
import { wrapFetchWithPayment, x402Client } from "@x402/fetch";
import { registerExactEvmScheme } from "@x402/evm/exact/client";
import { privateKeyToAccount } from "viem/accounts";

// Networks you will pay on, in order of preference. Default: Base. Set CROSSCHECK_NETWORKS=eip155:84532 for free test USDC.
const NETWORKS = (process.env.CROSSCHECK_NETWORKS ?? "eip155:8453").split(",").map((n) => n.trim());
const MAX_ATOMIC = 100_000n; // $0.10 (USDC has 6 decimals)

export function payingFetch(): typeof fetch {
  const key = process.env.CROSSCHECK_WALLET_KEY;
  if (!key) throw new Error("Set CROSSCHECK_WALLET_KEY to the private key of a dedicated wallet");
  const pick = <T extends { network: string }>(options: T[]) => NETWORKS.map((n) => options.find((o) => o.network === n)).find(Boolean);
  const client = new x402Client((_version, options) => pick(options) ?? options[0]!);
  registerExactEvmScheme(client, { signer: privateKeyToAccount(key as `0x${string}`) });
  client.onBeforePaymentCreation(async ({ selectedRequirements: r }) => {
    if (!NETWORKS.includes(r.network)) return { abort: true, reason: `crosscheck does not offer ${NETWORKS.join(", ")}` };
    if (BigInt(r.amount) > MAX_ATOMIC) return { abort: true, reason: `price ${r.amount} is above the limit` };
    return undefined;
  });
  return wrapFetchWithPayment(fetch, client) as typeof fetch;
}

/** POST a paid request and wait for the result (a 202 means paid and queued). */
export async function postAndWait(url: string, body: unknown): Promise<Record<string, any>> {
  const res = await payingFetch()(url, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body) });
  let out = (await res.json()) as Record<string, any>;
  if (res.status !== 200 && res.status !== 202) throw new Error(`crosscheck returned ${res.status}: ${JSON.stringify(out)}`);
  const { result_url, result_token } = out;
  while (out.status === "pending") {
    await new Promise((r) => setTimeout(r, (out.retry_after_seconds ?? 5) * 1000));
    out = await (await fetch(result_url, { headers: { authorization: `Bearer ${result_token}` } })).json();
  }
  return out;
}
